### PR TITLE
Consider ARAs in 2P

### PR DESCRIPTION
**Pull request by Godelaine**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
fix

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

   1) In post 2P sensi : apply ARAs from 2nd auto only
    2) Disable HVDC Range actions' AngleDroopActivePowerControl and set their setpoint to previously optimized setpoint
    3) Include auto and curative range actions in second preventive optimization : 
-  auto range actions : include for both global and non global scenarii
-  for non global 2P : exclude from optimization preventive and auto RA
   4) minor fix : include failed RA in 2P appliedRA
5) Bug fix : AFTER PRA flow/margin for curative cnecs post 2P optimization in RaoResult was wrong : 2P search tree optimization result was written, i.e preventive and curative RA were taken into account. Moreover, RaoResult now only contains instants for which we have a sensi computation result. For example, a curative cnec with 2P optimization does not have an AFTER ARA field.

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*
Flow/Margin after ARA for curative cnecs no longer available

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: sensitivity-analysis/src/main/java/com/farao_community/farao/sensitivity_analysis/SystematicSensitivityAdapter.java [hunk 1]
```diff
@@ -112,6 +112,9 @@ static SystematicSensitivityResult runSensitivity(Network network,
             network.getVariantManager().cloneVariant(workingVariantId, variantForState);
             network.getVariantManager().setWorkingVariant(variantForState);
 
+            // If some applied range actions are HvdcRangeActions, disable HvdcAngleDroopActivePowerControl and apply their optimal setpoint
+            findAndDisableHvdcAngleDroopActivePowerControl(network, appliedRemedialActions, state);
```
Comment (pet-mit) [inline sensitivity-analysis/src/main/java/com/farao_community/farao/sensitivity_analysis/SystematicSensitivityAdapter.java:116]: do this inside HvdcRangeAction.apply instead

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/castor/algorithm/CastorFullOptimization.java [hunk 2]
```diff
@@ -521,13 +519,13 @@ private static class SecondPreventiveRaoResult {
...
-        private final AppliedRemedialActions appliedCras;
+        private final AppliedRemedialActions appliedArasAndCras;
 
-        public SecondPreventiveRaoResult(PerimeterResult perimeterResult, PrePerimeterResult postPraSensitivityAnalysisOutput, Set<RemedialAction<?>> remedialActionsExcluded, AppliedRemedialActions appliedCras) {
+        public SecondPreventiveRaoResult(PerimeterResult perimeterResult, PrePerimeterResult postPraSensitivityAnalysisOutput, Set<RemedialAction<?>> remedialActionsExcluded, AppliedRemedialActions appliedArasAndCras) {
             this.perimeterResult = perimeterResult;
             this.postPraSensitivityAnalysisOutput = postPraSensitivityAnalysisOutput;
             this.remedialActionsExcluded = remedialActionsExcluded;
-            this.appliedCras = appliedCras;
+            this.appliedArasAndCras = appliedArasAndCras;
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/castor/algorithm/CastorFullOptimization.java:528]: is there a more elegant way to do all this ? i find it hard to follow.
maybe remove this field and re-build the set everytime it is necessary instead of storing it temporarily then copying only some values ;
maybe add the Instant to addAppliedNetworkActionsPostContingency too, even if we call it twice...

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/castor/algorithm/CastorFullOptimization.java [hunk 3]
```diff
@@ -579,29 +579,27 @@ private SecondPreventiveRaoResult runSecondPreventiveRao(RaoInput raoInput,
...
             BUSINESS_LOGS.error("Systematic sensitivity analysis after preventive remedial actions after second preventive optimization failed");
         }
         BUSINESS_LOGS.info("----- Second preventive perimeter optimization [end]");
-        return new SecondPreventiveRaoResult(secondPreventiveResult, postPraSensitivityAnalysisOutput, remedialActionsExcluded, appliedCras);
+        return new SecondPreventiveRaoResult(secondPreventiveResult, postPraSensitivityAnalysisOutput, remedialActionsExcluded, appliedArasAndCras);
     }
 
     static void addAppliedNetworkActionsPostContingency(AppliedRemedialActions appliedRemedialActions, Map<State, OptimizationResult> postContingencyResults) {
-        postContingencyResults.forEach((state, optimizationResult) -> {
-            if (optimizationResult.getSensitivityStatus(state) != ComputationStatus.FAILURE) {
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/castor/algorithm/CastorFullOptimization.java:595]: why did you remove the condition ?

**Review by pet-mit — CHANGES_REQUESTED** [review]

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/castor/algorithm/CastorFullOptimization.java (hunk 3 above)
Comment (Godelaine) [reply ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/castor/algorithm/CastorFullOptimization.java:595]: to keep failed perimeters in 2P

File: data/crac/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/HvdcRangeActionImplTest.java [hunk 4]
```diff
@@ -62,15 +68,20 @@ public void applyPositiveSetpoint() {
         HvdcRangeAction hvdcRa = hvdcRangeActionAdder.newRange().withMin(-5).withMax(10).add()
                 .add();
         hvdcRa.apply(network, 5);
+        hvdcRa.apply(networkWithAngleDroop, 6);
         assertEquals(5, hvdcRa.getCurrentSetpoint(network), 1e-6);
+        assertEquals(6, hvdcRa.getCurrentSetpoint(networkWithAngleDroop), 1e-6);
```
Comment (pet-mit) [inline data/crac/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/HvdcRangeActionImplTest.java:73]: also assert that HvdcAngleDroopActivePowerControl has been disabled

File: data/crac/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/HvdcRangeActionImplTest.java [hunk 5]
```diff
@@ -62,15 +68,20 @@ public void applyPositiveSetpoint() {
...
     }
 
     @Test
     public void applyNegativeSetpoint() {
         HvdcRangeAction hvdcRa = hvdcRangeActionAdder.newRange().withMin(-5).withMax(10).add()
                 .add();
         hvdcRa.apply(network, -3);
+        hvdcRa.apply(networkWithAngleDroop, -4);
         assertEquals(-3, hvdcRa.getCurrentSetpoint(network), 1e-6);
+        assertEquals(-4, hvdcRa.getCurrentSetpoint(networkWithAngleDroop), 1e-6);
```
Comment (pet-mit) [inline data/crac/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/HvdcRangeActionImplTest.java:84]: same

File: ra-optimisation/search-tree-rao/src/test/java/com/farao_community/farao/search_tree_rao/castor/algorithm/CastorFullOptimizationTest.java [hunk 6]
```diff
@@ -403,16 +403,16 @@ public void testIsRangeActionPreventive() {
     public void testIsRangeActionCurative() {
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/test/java/com/farao_community/farao/search_tree_rao/castor/algorithm/CastorFullOptimizationTest.java:403]: add test for ARAs

**Review by pet-mit — CHANGES_REQUESTED** [review]

**Review by pet-mit — APPROVED** [review]
